pkg/nsx/services/subnet/subnet.go

```diff
@@ -459,3 +486,273 @@ func (service *SubnetService) UpdateSubnetSet(ns string, vpcSubnets []*model.Vpc
+		return nil, err
+	}
+
+	return statusList, nil
+}
+
+// UpdateNSXSubnetCache updates the cache with the NSX subnet and status list
+func (service *SubnetService) UpdateNSXSubnetCache(associatedResource string, nsxSubnet *model.VpcSubnet, statusList []model.VpcSubnetStatus) {
+	service.nsxSubnetCacheMutex.Lock()
+	defer service.nsxSubnetCacheMutex.Unlock()
+
+	service.NSXSubnetCache[associatedResource] = struct {
+		Subnet     *model.VpcSubnet
+		StatusList []model.VpcSubnetStatus
+	}{
+		Subnet:     nsxSubnet,
+		StatusList: statusList,
+	}
+	log.Info("Updated NSX subnet cache", "AssociatedResource", associatedResource)
+}
+
+// RemoveSubnetFromCache removes a subnet from the NSXSubnetCache
+func (service *SubnetService) RemoveSubnetFromCache(associatedResource string, reason string) {
+	service.nsxSubnetCacheMutex.Lock()
+	defer service.nsxSubnetCacheMutex.Unlock()
+	delete(service.NSXSubnetCache, associatedResource)
+	log.Info("Removed Subnet from cache", "reason", reason, "AssociatedResource", associatedResource)
+}
+
+// AddSharedSubnetToResourceMap adds a shared subnet CR to the resource map
+func (service *SubnetService) AddSharedSubnetToResourceMap(associatedResource string, namespacedName types.NamespacedName) {
+	service.sharedSubnetResourceMapMutex.Lock()
+	defer service.sharedSubnetResourceMapMutex.Unlock()
+
+	// If the set doesn't exist for this associatedResource, create it
+	if _, exists := service.SharedSubnetResourceMap[associatedResource]; !exists {
+		service.SharedSubnetResourceMap[associatedResource] = sets.New[types.NamespacedName]()
+	}
+
+	// Add the namespacedName to the set (no need to check if it exists, sets handle that automatically)
+	service.SharedSubnetResourceMap[associatedResource].Insert(namespacedName)
+	log.Info("Added shared subnet to resource map", "AssociatedResource", associatedResource, "NamespacedName", namespacedName)
+}
+
+// RemoveSharedSubnetFromResourceMap removes a shared subnet CR from the resource map
+func (service *SubnetService) RemoveSharedSubnetFromResourceMap(associatedResource string, namespacedName types.NamespacedName) {
+	service.sharedSubnetResourceMapMutex.Lock()
+	defer service.sharedSubnetResourceMapMutex.Unlock()
+
+	// If the set exists for this associatedResource, remove the namespacedName
+	if namespacedNames, exists := service.SharedSubnetResourceMap[associatedResource]; exists {
+		namespacedNames.Delete(namespacedName)
+		log.Info("Removed shared subnet from resource map", "AssociatedResource", associatedResource, "NamespacedName", namespacedName)
+
+		// If the set is now empty, remove the associatedResource key
+		if namespacedNames.Len() == 0 {
+			delete(service.SharedSubnetResourceMap, associatedResource)
+		}
+	}
+}
+
+// GetSharedSubnetsForNamespace returns a map of associatedResource to namespaced names for a specific namespace
+func (service *SubnetService) GetSharedSubnetsForNamespace(namespace string) map[string][]types.NamespacedName {
```

Comment:
since in a namespace, 1:1 mapping for shared subnet with Subnet CR.
can we "return map[string]types.NamespacedName"?